File: Backend/src/Iit/Iit.Fibertest.Graph/Algorithms/LandmarksExtractors/ColoredLandmark.cs
using GMap.NET;
using Iit.Fibertest.Dto;

namespace Iit.Fibertest.Graph;

public class ColoredLandmark
{
    public bool IsFromBase;

    public Guid NodeId;
    public Guid FiberId; // to the left
    public int Number;
    public int NumberIncludingAdjustmentPoints;

    public string NodeTitle;
    public string NodeTitleColor;

    public string NodeComment;
    public string NodeCommentColor;

    public Guid EquipmentId;
    public string EquipmentTitle;
    public string EquipmentTitleColor;
    public EquipmentType EquipmentType;
    public string EquipmentTypeColor;

    public int LeftCableReserve;
    public int RightCableReserve;
    public string CableReservesColor;

    public double GpsDistance; // by GPS, ignore cable reserve
    public double GpsSection;
    public bool IsUserInput;
    public string GpsSectionColor;
    public double OpticalDistance; // from sor
    public double OpticalSection;
    public int EventNumber; // -1 если не связан

    public PointLatLng GpsCoors;
    public string GpsCoorsColor;

    public bool IsChanged { get => _isChanged(); }

    public ColoredLandmark FromLandmark(Landmark landmark, ColoredLandmark? oldLandmarkRow)
    {
        Number = landmark.Number;
        NumberIncludingAdjustmentPoints = landmark.NumberIncludingAdjustmentPoints;
        NodeId = landmark.NodeId;
        FiberId = landmark.FiberId;
        NodeTitle = landmark.NodeTitle;
        NodeTitleColor = oldLandmarkRow == null || landmark.NodeTitle == oldLandmarkRow.NodeTitle
            ? "Transparent" : "Cornsilk";
        NodeComment = landmark.NodeComment;
        NodeCommentColor = oldLandmarkRow == null || landmark.NodeComment == oldLandmarkRow.NodeComment
            ? "Transparent" : "Cornsilk";
        EquipmentId = landmark.EquipmentId;
        EquipmentTitle = landmark.EquipmentTitle;
        EquipmentTitleColor = oldLandmarkRow == null || landmark.EquipmentTitle == oldLandmarkRow.EquipmentTitle
            ? "Transparent" : "Cornsilk";
        EquipmentType = landmark.EquipmentType;
        EquipmentTypeColor = oldLandmarkRow == null || EquipmentType == oldLandmarkRow.EquipmentType
            ? "Transparent" : "Cornsilk";
        LeftCableReserve = landmark.LeftCableReserve;
        RightCableReserve = landmark.RightCableReserve;
        CableReservesColor = oldLandmarkRow == null ||
            (LeftCableReserve == oldLandmarkRow.LeftCableReserve && RightCableReserve == oldLandmarkRow.RightCableReserve)
            ? "Transparent" : "Cornsilk";

        GpsDistance = landmark.GpsDistance;
        GpsSection = landmark.GpsSection;
        IsUserInput = landmark.IsUserInput;
        GpsSectionColor = CalculateGpsSectionBrush(landmark, oldLandmarkRow);

        OpticalDistance = landmark.OpticalDistance;
        OpticalSection = landmark.OpticalSection;
        EventNumber = landmark.EventNumber;
        GpsCoors = landmark.GpsCoors;
        GpsCoorsColor = oldLandmarkRow == null || landmark.GpsCoors.Equals(oldLandmarkRow.GpsCoors)
            ? "Transparent" : "Cornsilk";

        return this;
    }

    public void ToLandmark(Landmark ll)
    {
        ll.Number = this.Number;
        ll.NumberIncludingAdjustmentPoints = this.NumberIncludingAdjustmentPoints;
        ll.NodeId = this.NodeId;
        ll.FiberId = this.FiberId;
        ll.NodeTitle = this.NodeTitle;
        ll.NodeComment = this.NodeComment;
        ll.EquipmentId = this.EquipmentId;
        ll.EquipmentTitle = this.EquipmentTitle;
        ll.EquipmentType = this.EquipmentType;
        ll.LeftCableReserve = this.LeftCableReserve;
        ll.RightCableReserve = this.RightCableReserve;
        ll.GpsDistance = this.GpsDistance;
        ll.GpsSection = this.GpsSection;
        ll.IsUserInput = this.IsUserInput;
        ll.OpticalDistance = this.OpticalDistance;
        ll.OpticalSection = this.OpticalSection;
        ll.EventNumber = this.EventNumber;
        ll.GpsCoors = this.GpsCoors;
    }

    private bool _isChanged()
    {
        return NodeCommentColor == "Cornsilk" || NodeTitleColor == "Cornsilk" 
               || EquipmentTitleColor == "Cornsilk" || EquipmentTypeColor == "Cornsilk"
               || CableReservesColor == "Cornsilk" || GpsSectionColor == "Cornsilk"|| GpsCoorsColor == "Cornsilk";
    }


    private string CalculateGpsSectionBrush(Landmark source, ColoredLandmark? oldLandmarkRow)
    {
        if (oldLandmarkRow != null)
        {
            if (oldLandmarkRow.IsUserInput ^ source.IsUserInput) 
                return "Cornsilk";
            if (Math.Abs(GpsSection - oldLandmarkRow.GpsSection) > 0.001)
                return "Cornsilk";
        }

        return  source.IsUserInput ? "LightGray" : "Transparent";
    }

}

File: WebFt30/src/app/core/store/models/ft30/colored-landmark.ts
import { EquipmentType } from 'src/grpc-generated';

export class LandmarksModel {
  landmarksModelId!: string;
  landmarks!: ColoredLandmark[];

  constructor(init?: Partial<LandmarksModel>) {
    Object.assign(this, init);
  }

  clone(): LandmarksModel {
    const clone = new LandmarksModel();
    clone.landmarksModelId = this.landmarksModelId;
    clone.landmarks = this.landmarks.map((l) => l.clone());
    return clone;
  }
}

export class ColoredLandmark {
  isFromBase!: boolean;

  nodeId!: string;
  fiberId!: string; // to the left
  number!: number;
  numberIncludingAdjustmentPoints!: number;

  nodeTitle!: string;
  nodeTitleColor!: string;

  nodeComment!: string;
  nodeCommentColor!: string;

  equipmentId!: string;
  equipmentTitle!: string;
  equipmentTitleColor!: string;
  equipmentType!: EquipmentType;
  equipmentTypeColor!: string;

  leftCableReserve!: number;
  rightCableReserve!: number;
  cableReservesColor!: string;

  gpsDistance!: number; // by GPS, ignore cable reserve
  gpsSection!: number;
  isUserInput!: boolean;
  gpsSectionColor!: string;

  opticalDistance!: number; // from sor
  opticalSection!: number;
  eventNumber!: number;

  gpsCoors!: L.LatLng;
  gpsCoorsColor!: string;

  public isSelected = false;

  public get UserInputLength() {
    return this.isUserInput ? this.gpsSection * 1000 : 0;
  }

  public set UserInputLength(value: number) {
    if (value === 0) {
      this.isUserInput = false;
      this.gpsSection = -1;
    } else {
      this.isUserInput = true;
      this.gpsSection = value / 1000;
    }
  }

  public clone(): ColoredLandmark {
    const clone = new ColoredLandmark();
    Object.assign(clone, this);
    return clone;
  }

  public cableReserveString(): string {
    if (this.equipmentType === EquipmentType.CableReserve) return this.leftCableReserve.toString();

    if (
      this.equipmentType === EquipmentType.Closure ||
      this.equipmentType === EquipmentType.Cross ||
      this.equipmentType === EquipmentType.Other ||
      this.equipmentType === EquipmentType.Terminal
    )
      return `${this.leftCableReserve} / ${this.rightCableReserve}`;

    return '';
  }

  public isEquipmentLandmark(): boolean {
    return (
      this.equipmentType === EquipmentType.Rtu ||
      this.equipmentType === EquipmentType.CableReserve ||
      this.equipmentType === EquipmentType.Closure ||
      this.equipmentType === EquipmentType.Cross ||
      this.equipmentType === EquipmentType.Other ||
      this.equipmentType === EquipmentType.Terminal
    );
  }

  public eventNumberString(): string {
    if (this.eventNumber === -1) return 'i18n.ft.no';
    return this.eventNumber.toString();
  }

  public canChangeEquipment(): boolean {
    return (
      this.equipmentType !== EquipmentType.Rtu && this.equipmentType !== EquipmentType.EmptyNode
    );
  }

  public areNodePropertiesChanged(other: ColoredLandmark): boolean {
    return (
      this.nodeTitle !== other.nodeTitle ||
      this.nodeComment !== other.nodeComment ||
      !this.areGpsCoorsEqual(other.gpsCoors, 0.0000005)
    );
  }

  private areGpsCoorsEqual(otherGpsCoors: L.LatLng, tolerance: number): boolean {
    const equals =
      Math.abs(this.gpsCoors.lat - otherGpsCoors.lat) < tolerance &&
      Math.abs(this.gpsCoors.lng - otherGpsCoors.lng) < tolerance;
    return equals;
  }

  public areEquipmentPropertiesChanged(other: ColoredLandmark): boolean {
    return (
      this.equipmentTitle !== other.equipmentTitle ||
      this.equipmentType !== other.equipmentType ||
      this.leftCableReserve !== other.leftCableReserve ||
      this.rightCableReserve !== other.rightCableReserve
    );
  }

  public areAnyPropertyChanged(other: ColoredLandmark): boolean {
    return (
      this.areNodePropertiesChanged(other) ||
      this.areEquipmentPropertiesChanged(other) ||
      this.UserInputLength !== other.UserInputLength
    );
  }
}
